Angular/angular-mentoring/src/app/course-list/course-list.component.ts

```diff
@@ -24,28 +27,34 @@ export class CourseListComponent implements OnInit {
 
   public courses: CourseItem[];
 
+  public courses$ = this.store.pipe(select(selectCourses));
+
   public params: string;
 
   public loading: boolean = false;
 
-  constructor(private coursesService: CoursesService, public dialog: MatDialog, private route: ActivatedRoute) { }
+  constructor(
+    private coursesService: CoursesService, 
+    private store: Store, 
+    public dialog: MatDialog, 
+    private route: ActivatedRoute
+  ) { }
 
   ngOnInit(): void {
     this.loadCourses();
+
     this.searchFilterSubject
       .pipe(debounceTime(250))
       .subscribe((val) => this.loadCourses(val));
+
     this.params = this.route.snapshot.params.id;
 
     if (this.params) this.openNewCourseModal(this.params);
   }
 
   public loadCourses(filter: string = ''): void {
     this.loading = true;
-    this.coursesService.getCourses(filter).subscribe(response => {
-      this.courses = response;
-      this.loading = false;
-    });
+    this.coursesService.getCourses(filter).subscribe((course) => this.store.dispatch(retrievedCoursesList( { CourseItem: course })));
```

Comment:
Ideally we would assign the service to dispatch the state.

Then the components only subscribe to the Store state.

Then where to call the `coursesService.getCourses`? I would suggest implementing a route resolver, which would call the service and load the page only after the data is acquired. But right now, in the way it is, it's alright.